### PR TITLE
Extension field

### PR DESCRIPTION
**Pull request by lonerapier**
This PR closes issue #40 

It changes the following:

- adds 2 degree extension field for `PlutoField`. I've not made field extension generic over degree. let me know if that is required.

### CODE REVIEW

File: src/field/gf_101_2.rs [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+};
+use std::ops::{Div, DivAssign, MulAssign, Rem};
+
+use serde::{Deserialize, Serialize};
+
+use super::ExtensionField;
+use crate::field::FiniteField;
+
+/// Pluto curve with modulus 101 supports two degree extension field. This can be verified
+/// by finding out embedding degree of the curve, i.e. smallest k: r|q^k-1
```
Comment (0xJepsen) [inline src/field/gf_101_2.rs:13]: When reading about the embedding degree I was trying to see this by hand but came across some questions. In our case r is our curve order which is 17 and p is field order which is 101. How does 17 | 101 ^2 - 1 which is 17 | 10200. 17 doesn't divide 10200. Also 10200 mod 101 is just 17 | 100. Do I have an error in my arithmetic? Are we working in a field here?

File: src/field/gf_101_2.rs [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+const EXT_DEGREE: usize = 2;
+const QUADRATIC_EXTENSION_FIELD_ORDER: u32 = 101 * 101;
+
+#[derive(Clone, Default, Copy, Eq, PartialEq, Hash, Debug, Serialize, Deserialize)]
+
+/// Quadratic Extension field element represented as polynomial of degree 1 in form:
+/// a_0 + a_1*t where {a_0, a_1} \in \mathhbb{F}. Uses irreducible poly of the form:
```
Comment (0xJepsen) [inline src/field/gf_101_2.rs:20]: I think this is the right approach but curious how or if we should use this in our polynomial module. I know @Autoparallel you have been working on the FFTs and polynomial interface. Maybe you have some insight here?

File: src/field/gf_101_2.rs [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+/// (X^2-K).
```
Comment (0xJepsen) [inline src/field/gf_101_2.rs:21]: should this be X^2 -K or X^2 + K?  X^2 + 2 is irreducible in F_{101} and should be used to generate our extension field.

File: src/field/gf_101_2.rs [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+    let scalar =
+      (self.value[0].square() - Self::irreducible() * self.value[1].square()).inverse().unwrap();
+    res.value[0] = self.value[0] * scalar;
+    res.value[1] = -self.value[1] * scalar;
+    Some(res)
+  }
+
+  #[inline]
+  fn pow(&self, power: Self::Storage) -> Self {
+    // TODO: not ideal to duplicate across traits impls
```
Comment (0xJepsen) [inline src/field/gf_101_2.rs:90]: Maybe we can add a default impl in the trait what do you think?

File: src/field/gf_101.rs [hunk 5]
```diff
@@ -83,6 +83,8 @@ impl FiniteField for GF101 {
   fn neg_one() -> Self { Self::new(Self::ORDER - 1) }
 
   fn generator() -> Self { Self::new(2) }
+
+  fn from_canonical_u32(n: u32) -> Self { Self::new(n) } // TODO: recheck
```
Comment (Autoparallel) [inline src/field/gf_101.rs:87]: This is a nit, but we don't really need this if we just use `GF101::new()` right?

File: src/field/gf_101_2.rs (hunk 2 above)
Comment (Autoparallel) [reply src/field/gf_101_2.rs:20]: Let me get the polynomial PR up.

File: src/field/gf_101_2.rs (hunk 3 above)
Comment (Autoparallel) [reply src/field/gf_101_2.rs:21]: I think both work as (correct me if I'm wrong) 2 is not a square in GF_{101}.

File: src/field/gf_101_2.rs (hunk 4 above)
Comment (Autoparallel) [reply src/field/gf_101_2.rs:90]: Looking good!

**Review by Autoparallel — APPROVED** [review]
I'm good with this to go in as is :)

File: src/field/gf_101_2.rs (hunk 3 above)
Comment (lonerapier) [reply src/field/gf_101_2.rs:21]: yeah, both of them are irreducible, added it in the field sage script as well

https://github.com/pluto/ronkathon/blob/9c0ba7055b0f80b0d157243ca48bf002ed823ca4/math/field.sage#L47-L51

File: src/field/gf_101.rs (hunk 5 above)
Comment (lonerapier) [reply src/field/gf_101.rs:87]: but then that's struct specific function right? We need a trait specific function to create new values, for example: in extension fields, each subvalue in extension field element is base field element.

https://github.com/pluto/ronkathon/blob/9c0ba7055b0f80b0d157243ca48bf002ed823ca4/src/field/gf_101_2.rs#L69

Is there a better way in Rust to do this? Sorry, still in my learning path for Rust.

File: src/field/gf_101_2.rs (hunk 4 above)
Comment (lonerapier) [reply src/field/gf_101_2.rs]: done

**Comment by lonerapier** [comment]
@0xJepsen @Autoparallel changes done. let me know if any others are needed

File: src/field/gf_101_2.rs (hunk 1 above)
Comment (0xJepsen) [reply src/field/gf_101_2.rs:13]: that's a classic, my bad.

**Pull request merged**